lightning/src/ln/channelmanager.rs

```diff
@@ -3103,20 +3169,20 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 
 	fn claim_funds_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder<Signer>>, source: HTLCSource, payment_preimage: PaymentPreimage, forwarded_htlc_value_msat: Option<u64>, from_onchain: bool) {
 		match source {
-			HTLCSource::OutboundRoute { session_priv, .. } => {
+			HTLCSource::OutboundRoute { session_priv, mpp_id, .. } => {
 				mem::drop(channel_state_lock);
-				if {
-					let mut session_priv_bytes = [0; 32];
-					session_priv_bytes.copy_from_slice(&session_priv[..]);
-					self.pending_outbound_payments.lock().unwrap().remove(&session_priv_bytes)
-				} {
-					let mut pending_events = self.pending_events.lock().unwrap();
-					pending_events.push(events::Event::PaymentSent {
-						payment_preimage
-					});
-				} else {
-					log_trace!(self.logger, "Received duplicative fulfill for HTLC with payment_preimage {}", log_bytes!(payment_preimage.0));
+				let mut session_priv_bytes = [0; 32];
+				session_priv_bytes.copy_from_slice(&session_priv[..]);
+				let mut outbounds = self.pending_outbound_payments.lock().unwrap();
+				if let Some(mut sessions) = outbounds.remove(&mpp_id) {
+					if sessions.remove(&session_priv_bytes) {
+						self.pending_events.lock().unwrap().push(
+							events::Event::PaymentSent { payment_preimage }
+						);
+						return
```

Comment:
Do we not want to drop the whole pending payment at this point? If we get one preimage, we should consider the payment success, irrespective of future HTLC failures.

Reply:
I believe that's how it's implemented: `outbounds.remove(&mpp_id)` above

Reply:
Ah, duh, yes. Next question, should we not just always generate `PaymentSent` even if we've already removed the session priv? Or, really, isn't that unreachable? I'm not sure how we could hit the first if but not the second.

Reply:
Made it unreachable!

Reply:
Wait, I think this is reachable if we receive a duplicate `update_fulfill_htlc` msg? So I think the right response is to log that there's been a duplicate, not be `unreachable`

Reply:
Ah, you mean if we sent two payments to the same payment hash and have a duplicate claim on the first? I believe you may be correct.